### PR TITLE
New Class for Conditioned Spatial Random Fields

### PR DESCRIPTION
**Pull request by MuellerSeb**
Closes: #128, #126 

This PR adds a new class for conditioned random fields `CondSRF`.

This class replaces the conditioning feature of the SRF class, which was limited to Ordinary and Simple kriging.

`CondSRF` behaves similar to the `SRF` class, but instead of a covariance model, it takes a kriging class as input. With this kriging class, all conditioning related settings are defined.

During refactoring the `RandMeth` class, a switch to select the sampling strategy was added. 

# Note
This change breaks backward compatibility, since conditioned random fields are no longer provided by the `SRF` class.

### CODE REVIEW

**Comment by MuellerSeb** [comment]
@LSchueler this is ready for review. :relaxed:

File: gstools/field/cond_srf.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+import numpy as np
+from gstools.field.generator import RandMeth
+from gstools.field.base import Field
+from gstools.krige import Krige
+
+__all__ = ["CondSRF"]
+
+GENERATOR = {
+    "RandMeth": RandMeth,
+}
```
Comment (LSchueler) [inline gstools/field/cond_srf.py:23]: Maybe add a comment stating that this is the dict. of supported generators?

File: gstools/field/cond_srf.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        Name of the field generator to be used.
+        At the moment, the following generators are provided:
+
+            * "RandMeth" : The Randomization Method.
+              See: :any:`RandMeth`
+            * "IncomprRandMeth" : The incompressible Randomization Method.
+              This is the original algorithm proposed by Kraichnan 1970
+              See: :any:`IncomprRandMeth`
+            * "VectorField" : an alias for "IncomprRandMeth"
+            * "VelocityField" : an alias for "IncomprRandMeth"
```
Comment (LSchueler) [inline gstools/field/cond_srf.py:43]: These should not be listed here.

File: gstools/krige/base.py [hunk 3]
```diff
@@ -16,7 +16,7 @@
 from scipy.spatial.distance import cdist
 import scipy.linalg as spl
 from gstools.field.base import Field
-from gstools.krige.krigesum import krigesum
+from gstools.krige.krigesum import krigesum, krigesumfast
```
Comment (LSchueler) [inline gstools/krige/base.py:19]: `krigesum_fast`?

File: gstools/krige/krigesum.pyx [hunk 4]
```diff
@@ -37,3 +37,28 @@ def krigesum(
             field[k] += cond[i] * krig_fac
 
     return np.asarray(field), np.asarray(error)
+
+
+def krigesumfast(
```
Comment (LSchueler) [inline gstools/krige/krigesum.pyx:42]: I think a short comment on what the difference to `krigesum` is should be written here.
Have you measured the difference? - Maybe it's more `krigesumwithlessmemoryconsumption`? ;-)

File: gstools/random/rng.py [hunk 5]
```diff
@@ -61,10 +61,10 @@ def sample_ln_pdf(
         nwalkers : :class:`int`, optional
             The number of walkers in the mcmc sampler. Used for the
             emcee.EnsembleSampler class.
-            Default: 100
+            Default: 50
         burn_in : :class:`int`, optional
             Number of burn-in runs in the mcmc algorithm.
-            Default: 100
+            Default: 20
         oversampling_factor : :class:`int`, optional
```
Comment (LSchueler) [inline gstools/random/rng.py:68]: It's funny how some topics overlap ;-)

File: gstools/random/tools.py [hunk 6]
```diff
@@ -46,13 +46,9 @@ def seed(self):
...
         return self._seed
 
-    def __str__(self):
-        """Return String representation."""
-        return self.__repr__()
-
     def __repr__(self):
         """Return String representation."""
-        return "RNG(seed={})".format(self.seed)
+        return "MasterRNG(seed={})".format(self.seed)
```
Comment (LSchueler) [inline gstools/random/tools.py:51]: Good catch!

**Review by LSchueler — CHANGES_REQUESTED** [review]
Wow, the `SRF` class is looking soo much nicer now!

Really nice work!

I suggest to import `gstools.field.base.Field` to `gstools.field`. The import in one of the examples looks a but ugly.

I just made a few small remarks in the code. I think after you have addressed we are ready to go!!

File: gstools/random/rng.py (hunk 5 above)
Comment (MuellerSeb) [reply gstools/random/rng.py:68]: I couldn't resist...

File: gstools/krige/krigesum.pyx (hunk 4 above)
Comment (MuellerSeb) [reply gstools/krige/krigesum.pyx:42]: I am also not happy. Got a better name/idea?

File: gstools/field/cond_srf.py (hunk 2 above)
Comment (MuellerSeb) [reply gstools/field/cond_srf.py:43]: Good catch!

File: gstools/krige/krigesum.pyx (hunk 4 above)
Comment (LSchueler) [reply gstools/krige/krigesum.pyx:42]: Hmm, what about something in the direction of `calc_krige_and_error_field` and `calc_krige_field`?

I was just thinking, what about combining both functions with a bool `error` or `error_field` or `calc_error_field` as an argument? - I mean, with predictive branching, it shouldn't be much slower and with `error = False`, the array could be set to `cdef double[:] error = np.zeros(1)` or something for saving memory.
I guess that would have to be profiled to see if it actually is fast.
But that's also something very low on the priority list.

File: examples/06_conditioned_fields/00_condition_ensemble.py [hunk 7]
```diff
@@ -16,30 +16,35 @@
...
+#
+# Here we use a gaussian covariance model and ordinary kriging for conditioning
+# the spatial random field.
 
-# spatial random field class
 model = gs.Gaussian(dim=1, var=0.5, len_scale=1.5)
-srf = gs.SRF(model)
-srf.set_condition(cond_pos, cond_val, "ordinary")
+krige = gs.krige.Ordinary(model, cond_pos, cond_val)
+csrf = gs.CondSRF(krige)
```
[the remark this answers is not in the document]
Comment (MuellerSeb) [reply examples/06_conditioned_fields/00_condition_ensemble.py]: Done.

File: examples/06_conditioned_fields/00_condition_ensemble.py [hunk 8]
```diff
@@ -16,30 +16,35 @@
...
 
 ###############################################################################
 
 fields = []
 for i in range(100):
     # print(i) if i % 10 == 0 else None
```
[the remark this answers is not in the document]
Comment (MuellerSeb) [reply examples/06_conditioned_fields/00_condition_ensemble.py]: Done.

File: gstools/field/cond_srf.py (hunk 1 above)
Comment (MuellerSeb) [reply gstools/field/cond_srf.py:23]: Done.

File: gstools/field/cond_srf.py (hunk 2 above)
Comment (MuellerSeb) [reply gstools/field/cond_srf.py]: Done.

File: gstools/krige/base.py (hunk 3 above)
Comment (MuellerSeb) [reply gstools/krige/base.py]: Renamed to `calc_field_krige_and_variance` and `calc_field_krige`.

File: gstools/krige/krigesum.pyx (hunk 4 above)
Comment (MuellerSeb) [reply gstools/krige/krigesum.pyx]: Renamed to `calc_field_krige_and_variance` and `calc_field_krige`. Added method `_summate` to `Krige` to care about selection of the correct routine.

**Review by LSchueler — APPROVED** [review]
:-) Found another typo ;-)

But anyway, here is a big thumbs up!

**Pull request merged**